metals/src/main/scala/scala/meta/internal/metals/onTypeRangeFormatters/IndentOnPaste.scala

```diff
@@ -0,0 +1,173 @@
+import scala.util.matching.Regex
+
+object IndentOnPaste extends RangeFormatter {
+
+  val increaseIndentPatternRegex: Regex =
+    raw"""(((<!\bend\b\s*?)\b(if|while|for|match|try))|(\bif\s+(?!.*?\bthen\b.*?$$)[^\s]*?)|(\b(then|else|do|catch|finally|yield|case))|=|=>|<-|=>>|:)\s*?$$|(^.*\{[^}"']*$$)""".r
+  val indentRegex: Regex = raw"\S".r
+
+  private def codeStartPosition(line: String): Option[Int] =
+    indentRegex.findFirstMatchIn(line).map(_.start)
+
+  private def increaseIndentation(line: String) =
+    increaseIndentPatternRegex.findFirstIn(line).nonEmpty
+
+  private def stringRepeat(s: String, n: Int): String =
+    ("%0" + n + "d").format(0).replace("0", s)
+
+  // converts spaces into tabs and vice-versa, normalizing the lengths of indentations
+  private def normalizeSpacesAndTabs(
+      line: String,
+      codeStartPos: Int,
+      blank: String,
+      tabSize: Int
+  ): String = {
+    if (codeStartPos != 0) {
+      val substrLength = math.min(line.length, codeStartPos)
+      val indentation = line.substring(0, substrLength)
+      val indentChars = indentation.split("")
+      val pastedBlank = indentChars.head
+      blank match {
+        case "\t" if pastedBlank == blank => line
+        case " " if pastedBlank == blank => line
+        case "\t" if pastedBlank == " " =>
+          val tabNum = math.ceil(pastedBlank.length / 2).toInt
+          stringRepeat(blank, tabNum) ++ line.slice(codeStartPos, line.length)
+        case " " if pastedBlank == "\t" =>
+          stringRepeat(blank, tabSize) ++ line.slice(
+            codeStartPos,
+            line.length
+          )
+        case _ => line
+      }
+    } else line
+  }
+
+  override def contribute(
+      sourceText: String,
+      range: lsp4j.Range,
+      splitLines: Array[String],
+      startPos: StartPosition,
+      endPos: EndPosition,
+      formattingOptions: FormattingOptions,
+      tokens: Option[Tokens]
+  ): Option[List[TextEdit]] = {
+    val insertSpaces = formattingOptions.isInsertSpaces
+    val originalTabSize = formattingOptions.getTabSize
+    val rangeStart = startPos.toLSP.getStart
+    rangeStart.setCharacter(0)
+    val pastedRange = new Range(rangeStart, endPos.toLSP.getEnd)
+    val startLine = startPos.toLSP.getStart.getLine
+    val endLine = endPos.toLSP.getEnd.getLine
+
+    val splitLinesWithIndex = splitLines.zipWithIndex
+    val inRangeLines = splitLinesWithIndex.slice(startLine, endLine + 1)
+    val pastedLines = inRangeLines.map(_._1)
+    val pastedLinesWithIndex = pastedLines.zipWithIndex
+
+    val (blank, tabSize) =
+      if (insertSpaces) (" ", originalTabSize) else ("\t", 1)
+
+    // These are the lines from the first pasted line, going above
+    val prePastedLines = splitLinesWithIndex.take(startLine).reverse
+
+    val currentIndentationLevel = (for {
+      (line, _) <- prePastedLines.find(t => {
+        val trimmed = t._1.trim()
+        trimmed.nonEmpty && !trimmed.startsWith("|")
```

Comment:
Essentially it's to support this case:

> Anyone that can help me find unsupported edge cases that I did not got right, is welcome!
> An example of such edge case is the when parsing after a multi-line string:
> 
> ```scala
> object Main:
>   val myMultilineString: String = """| My
>                                      | string""".stripMargin
>   //@@ pasting here
> end Main
> ```
> 
> the correct indentation found to indent the first line should not be considered the one of the piped line (` | string`) but the one of the actual string declaration (`val myMultilineString ...`)

Reply:
Ach yeah! After thinking a bit I realized the reason for it. I wonder though if we would be able to cover all possible cases. Since you can also have for example (I think that is only in Scala 3):
```scala
  val intNum = 123 
    + 123
```

Not sure what other possibilities we can have.